### PR TITLE
Move Device members from public to private

### PR DESCRIPTION
**Pull request by ayerofieiev-tt**
### Ticket
None

### Problem description
Most members of class `Device` are currently exposed as public. 
This allows anyone to reach into internals and prevents building a clean interface.

### What's changed
This PR does not intend to fix the class design, but only makes a small step forward:
* Most members moved to private
* New getter/setter functions were added as necessary 
* All private members have name with `_` at the end, aligned with an established codestyle
* Removed templates where not needed (e.g. simply return DeviceAddr)
* Moved minor impl to .cpp from .hpp
* Default initialized primitive members 

Next steps:
* Move some of member functions to private
* Move more of impl from .hpp to .cpp
* Remove some of the templates (e.g. get_dev_addr)

### Checklist
- [x] [Post commit CI](https://github.com/tenstorrent/tt-metal/actions/runs/12451981987)

### CODE REVIEW

File: tt_metal/impl/device/device.cpp [hunk 1]
```diff
@@ -3732,6 +3732,24 @@ std::shared_ptr<TraceBuffer> Device::get_trace(uint32_t tid) {
     return this->active_sub_device_manager_->get_trace(tid);
 }
 
+void Device::enable_program_cache() {
```
Comment (ayerofieiev-tt) [inline tt_metal/impl/device/device.cpp:3735]: moved from .hpp

File: tt_metal/impl/device/device.cpp [hunk 2]
```diff
@@ -3884,6 +3902,20 @@ const std::vector<SubDeviceId> &Device::get_sub_device_ids() const {
     return this->active_sub_device_manager_->get_sub_device_ids();
 }
 
+DeviceAddr Device::get_dev_addr(CoreCoord virtual_core, HalL1MemAddrType addr_type) const {
```
Comment (ayerofieiev-tt) [inline tt_metal/impl/device/device.cpp:3905]: moved from .hpp, removed template

**Review by cfjchu — APPROVED** [review]
glorious 🔥

File: tt_metal/impl/device/device.cpp [hunk 3]
```diff
@@ -3884,6 +3902,16 @@ const std::vector<SubDeviceId> &Device::get_sub_device_ids() const {
     return this->active_sub_device_manager_->get_sub_device_ids();
 }
 
+DeviceAddr Device::get_base_allocator_addr(const HalMemType &mem_type) const {
```
Comment (ayerofieiev-tt) [inline tt_metal/impl/device/device.cpp:3905]: moved from .hpp, removed template

File: tt_metal/impl/device/device.hpp [hunk 4]
```diff
@@ -292,84 +292,29 @@ class Device {
...
     void synchronize();
     void set_worker_mode(const WorkExecutorMode& mode);
     void enable_async(bool enable);
-    WorkExecutorMode get_worker_mode() { return work_executor.get_worker_mode(); }
-    void set_worker_queue_mode(const WorkerQueueMode& mode) { this->work_executor.set_worker_queue_mode(mode); }
-    WorkerQueueMode get_worker_queue_mode() { return this->work_executor.get_worker_queue_mode(); }
-    // TODO: Uplift usage of friends. Buffer and Program just need access to allocator
-    friend class Buffer;
-    friend class Program;
-    friend class SystemMemoryManager;
```
Comment (ayerofieiev-tt) [inline tt_metal/impl/device/device.hpp:306]: no more friends

**Review by jaykru-tt — APPROVED** [review]

File: tt_metal/impl/device/device.hpp [hunk 5]
```diff
@@ -396,12 +341,65 @@ class Device {
...
+private:
     void initialize_default_sub_device_state(size_t l1_small_size, size_t trace_region_size, tt::stl::Span<const std::uint32_t> l1_bank_remap);
     SubDeviceManagerId get_next_sub_device_manager_id();
     void reset_sub_devices_state(const std::unique_ptr<detail::SubDeviceManager>& sub_device_manager);
     void MarkAllocationsUnsafe();
     void MarkAllocationsSafe();
+
+    chip_id_t id_;
+    uint32_t build_key_ = 0;
+    std::map<uint32_t, std::map<chip_id_t, std::vector<std::vector<std::tuple<tt_cxy_pair, dispatch_worker_build_settings_t>>>>> tunnel_device_dispatch_workers_;
```
Comment (omilyutin-tt) [inline tt_metal/impl/device/device.hpp:370]: Woah, probably needs a named type...

**Review by omilyutin-tt — APPROVED** [review]
Great, thanks Artem!

File: tt_metal/impl/device/device.hpp (hunk 5 above)
Comment (dmakoviichuk-tt) [reply tt_metal/impl/device/device.hpp:370]: MapOfMapOfVectorOfVectorOfTuple ooh I never saw so large std type.
It should a class or map of classes. Type naming doesn't help here.

**Review by dmakoviichuk-tt — APPROVED** [review]

File: tt_metal/impl/device/device.hpp (hunk 5 above)
Comment (ayerofieiev-tt) [reply tt_metal/impl/device/device.hpp:370]: > Type naming doesn't help here. 

🤣

yes

**Review by aliuTT — APPROVED** [review]

**Review by TT-BrianLiu — APPROVED** [review]

**Pull request merged**